### PR TITLE
Added support for unlimited supplies in market shops

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: #6569

* **Server Mode**: Both

* **Description of Pull Request**: 
Added support for unlimited supplies in market shops.

### CODE REVIEW

**Comment by JohnnyPlayy** [comment]
You are awesome!

File: doc/script_commands.txt [hunk 1]
```diff
@@ -320,6 +320,9 @@ temporary character variables, permanent local account variables or permanent gl
 variables. These variables must be of integer type, not string. 'discount' flag is an
 optional value which makes the price at that shop become affected by discount skill.
 
+"marketshop" can have limited quantity of an item in stock.
+If you want to have unlimited stock in a marketshop please use -1 as stock.
```
Comment (aleos89) [inline doc/script_commands.txt:324]:
```suggestion
Use -1 in the stock field to have unlimited stock in a marketshop.
```

File: doc/script_commands.txt [hunk 2]
```diff
@@ -7263,12 +7267,13 @@ The function returns 1 if shop was updated successfully, or 0 if not found.
 
 NOTES:
  - That you cannot use -1 to specify default selling price!
```
Comment (aleos89) [inline doc/script_commands.txt:7269]:
```suggestion
 - A sell price of -1 is not a valid value!
```

File: doc/script_commands.txt [hunk 3]
```diff
@@ -7263,12 +7267,13 @@ The function returns 1 if shop was updated successfully, or 0 if not found.
...
- - If attached shop type is market shop, need an extra param after price, it's <qty>
-   and make sure don't add duplication item!
+ - If attached shop type is market shop, need an extra param after price, it's <stock>
+   and make sure don't add duplication item! For unlimited stock use -1.
```
Comment (aleos89) [inline doc/script_commands.txt:7271]:
```suggestion
 - If the attached shop type is a market shop, notice that there is an extra param after price, <stock>. Make sure to not add duplicate items! For unlimited stock use -1.
```

File: doc/script_commands.txt [hunk 4]
```diff
@@ -7278,8 +7283,8 @@ The function returns 1 if shop was updated successfully, or 0 if not found.
 
 NOTES:
  - That you cannot use -1 to specify default selling price!
```
Comment (aleos89) [inline doc/script_commands.txt:7285]:
```suggestion
 - A sell price of -1 is not a valid value!
```

File: doc/script_commands.txt [hunk 5]
```diff
@@ -7278,8 +7283,8 @@ The function returns 1 if shop was updated successfully, or 0 if not found.
...
- - If attached shop type is market shop, need an extra param after price, it's <qty>
-   and make sure don't add duplication item!
+ - If attached shop type is market shop, need an extra param after price, it's <stock>
+   and make sure don't add duplication item! For unlimited stock use -1.
```
Comment (aleos89) [inline doc/script_commands.txt:7287]:
```suggestion
 - If attached shop type is market shop, need an extra param after price, it's <stock>
   and make sure don't add duplication item! For unlimited stock use -1.
```

File: doc/script_commands.txt [hunk 6]
```diff
@@ -7319,7 +7324,7 @@ NOTES:
 
 Update an entry from shop. If price is 0 means don't change the price, maybe used for
 marketshop to update the stock quantity. Except marketshop type, 'stock' value means
```
Comment (aleos89) [inline doc/script_commands.txt:7326]:
```suggestion
Update an entry from a shop. If the price is 0 it won't be changed. May also be used for
marketshop to update the stock quantity. For unlimited stock, use -1.
```

File: doc/script_commands.txt [hunk 7]
```diff
@@ -7319,7 +7324,7 @@ NOTES:
...
-nothing.
+nothing. For unlimited stock use -1.
```
Comment (aleos89) [inline doc/script_commands.txt:7327]:
```suggestion
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -12319,8 +12321,22 @@ void clif_parse_NpcBuyListSend( int fd, struct map_session_data* sd ){
 
 	if( sd->state.trading || !sd->npc_shopid )
 		result = e_purchase_result::PURCHASE_FAIL_MONEY;
-	else
-		result = npc_buylist( sd, n, (struct s_npc_buy_list*)p->items );
+	else{
+		std::vector<s_npc_buy_list> items;
```
Comment (aleos89) [inline src/map/clif.cpp:12325]:
```suggestion
		std::vector<s_npc_buy_list> items = {};
```
Just following suite with how you declared it in `_buy`.

**Review by aleos89 — APPROVED** [review]
Just a few minor things, looks great!

**Comment by JohnnyPlayy** [comment]
https://user-images.githubusercontent.com/17951544/151638178-5d14cdde-588e-4fab-97d2-aa32da9ad68d.mp4


https://user-images.githubusercontent.com/17951544/151638205-1304bc12-e0b9-41f1-88a7-fa327cd4f811.mp4


It's working perfectly! Very good.
The only thing different is the Purchased Items Window that only opens once.
How much does KRO open whenever you buy something.

**Comment by Toshiro90** [comment]
> ```diff
> - - That you cannot use -1 to specify default selling price!
> + - A sell price of -1 is not a valid value!
> ```
Actually there's no check at all for negative values in `npcshopitem` and `npcshopadditem`.
grafik
When buying those items: `[Error]: pc_payzeny: Paying negative Zeny (zeny=-110, account_id=2000000, char_id=150001).`


> ```diff
> - Update an entry from shop. If price is 0 means don't change the price, maybe used for
> - marketshop to update the stock quantity. Except marketshop type, 'stock' value means
> + Update an entry from a shop. If the price is 0 it won't be changed. May also be used for
> + marketshop to update the stock quantity. For unlimited stock, use -1.
> ```
No negative price check in `npcshopupdate` as well.

Regarding consistency of above functions, while you can add free items (0 Zeny) with `npcshopitem` and `npcshopadditem`, it is not possible to change the price to 0 Zeny with `npcshopupdate`. Wouldn't it be more appropriate to use `-1` instead of `0` as special value for not changing the price when updating the stock amount?

File: doc/script_commands.txt (hunk 2 above)
Comment (Lemongrass3110) [reply doc/script_commands.txt:7269]: We could actually implement that?
As suggested from @Toshiro90

File: doc/script_commands.txt (hunk 4 above)
Comment (Lemongrass3110) [reply doc/script_commands.txt:7285]: Again - we could implement this.
As suggested by @Toshiro90

**Pull request merged**